torch/autograd/functional.py

```diff
@@ -0,0 +1,712 @@
+        result (tuple of Tensor or Tensor): result of the dot product with the same shape
+            as the inputs.
+
+    Example::
+
+        >>> def pow_reducer(x):
+        ...   return x.pow(3).sum()
+        >>> inputs = torch.rand(2, 2)
+        >>> v = torch.ones(2, 2)
+        >>> vhp(pow_reducer, inputs, v)
+       (tensor(0.5591),
+        tensor([[1.0689, 1.2431],
+                [3.0989, 4.4456]]))
+
+        >>> vhp(pow_reducer, inputs, v, create_graph=True)
+        (tensor(0.5591, grad_fn=<SumBackward0>),
+         tensor([[1.0689, 1.2431],
+                 [3.0989, 4.4456]], grad_fn=<MulBackward0>))
+
+
+        >>> def pow_adder_reducer(x, y):
+        ...   return (2 * x.pow(2) + 3 * y.pow(2)).sum()
+        >>> inputs = (torch.rand(2), torch.rand(2))
+        >>> v = (torch.zeros(2), torch.ones(2))
+        >>> vhp(pow_adder_reducer, inputs, v)
+        (tensor(4.8053),
+         (tensor([0., 0.]),
+          tensor([6., 6.])))
+
+    """
+
+    tuple_inputs, inputs = _as_tuple(inputs, "inputs", "vhp")
+    inputs = _grad_preprocess(inputs, create_graph=create_graph, need_graph=True)
+
+    if v is not None:
+        tuple_v, v = _as_tuple(v, "v", "vhp")
+        v = _grad_preprocess(v, create_graph=create_graph, need_graph=False)
+        _validate_v(v, inputs)
+    else:
+        if len(inputs) != 1 or inputs[0].nelement() != 1:
+            raise RuntimeError("The vector v can only be None if the input to the user-provided function "
+                               "is a single Tensor with a single element.")
+
+    outputs = func(*inputs)
+    tuple_outputs, outputs = _as_tuple(outputs, "outputs of the user-provided function", "vhp")
+    _check_requires_grad(outputs, "outputs", strict=strict)
+
+    if tuple_outputs or not torch.is_tensor(outputs[0]):
+        raise RuntimeError("The function given to vhp should return a single Tensor")
+
+    if outputs[0].nelement() != 1:
+        raise RuntimeError("The Tensor returned by the function given to vhp should contain a single element")
+
+    jac = _autograd_grad(outputs, inputs, create_graph=True)
+    _check_requires_grad(jac, "jacobian", strict=strict)
+
+    grad_res = _autograd_grad(jac, inputs, v, create_graph=create_graph)
+
+    vhp = tuple()
+    for i, vhp_i in enumerate(grad_res):
+        if vhp_i is None:
+            if strict:
+                raise RuntimeError("The jacobian of the user-provided function is independent of "
+                                   "input {}. This is not allowed in strict mode.".format(i))
+            vhp_i = torch.zeros_like(inputs[i])
+        else:
+            if strict and create_graph and not vhp_i.requires_grad:
+                raise RuntimeError("The hessian of the user-provided function is independent of "
+                                   "input {}. This is not allowed in strict mode when create_graph=True.".format(i))
+        vhp += (vhp_i,)
```

Comment:
This logic, post-processing None gradients from autograd.grad, seems common among all of the functions. Would it be worth it to toss this into a helper function?

The error messages look like they are going to need to be different so this might turn out to be a bad idea if the result of the move makes things less readable

Reply:
The error messages are different, but I guess I can use a flag for that.
Will refactor that into a single function 👍